### PR TITLE
Java: New itext7 Fathead

### LINKED ISSUE #624: Java: Create New iText 7 Community Fathead
# Create a new iText 7 Community Fathead Instant Answer
Help us make DuckDuckGo the best search engine for programmers!

## iText 7 Community

Here is where you will find our application programming interfaces (APIs) - the building blocks you may use to create your very own PDF solution.

http://developers.itextpdf.com/apis

### What do I need to know?
You'll need to know how to code in **Perl**, **Python**, **Ruby**, or **JavaScript**.
fathead languages

### What am I doing?
You will write a script that scrapes or downloads the data source below, and generates an **output.txt** file containing the parsed documentation. You can learn more about Fatheads and the `output.txt` syntax [**here**](https://docs.duckduckhack.com/resources/fathead-overview.html).

**Data source**: http://itextsupport.com/apidocs/itext7/7.0.0/

**Instant Answer Page**: https://duck.co/ia/view/itext7

### What is the Goal?
As part of our [Programming Mission](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53), we're aiming to reach 100% Instant Answer (IA) coverage for searches related to programming languages by creating new Instant Answers, and improving existing ones.

Here are some Fathead examples:

- Ruby Docs
    - [Code](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/ruby) | [Example Query](https://duckduckgo.com/?q=array+bsearch&ia=about)
- MDN CSS
    - [Code](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/mdn_css) | [Example Query](https://duckduckgo.com/?q=css+background-position&ia=about)

  css background-position search

[See more related Instant Answers](https://duck.co/ia?repo=fathead)

## Get Started
- [ ] 1) Claim this issue by commenting below
- [ ] 2) Review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/CONTRIBUTING.md)
- [ ] 3) [Set up your development environment](https://docs.duckduckhack.com/welcome/setup-dev-environment.html), and fork this repository
- [ ] 4) Create a new Instant Answer Page: https://duck.co/ia/new_ia (then let us know, here!)
- [ ] 5) Create the Fathead
- [ ] 6) Create a Pull Request
- [ ] 7) Ping @marti1125 for a review

## Resources
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) to ask questions
- Join the [DuckDuckHack Forum](https://forum.duckduckhack.com/) to discuss project planning and Instant Answer metrics
- Read the [DuckDuckHack Documentation](https://docs.duckduckhack.com/) for technical help

### PR DESCRIPTION
**Pull request by BillyBarbaro**
## Description of new Instant Answer, or changes

Creates the Instant Answer for the Java iText 7 Library.

## Related Issues and Discussions

Fixes #624 

## People to notify

@marti1125
@pjhampton 

## Testing & Review
To be completed by Language Leader (or DDG Staff) when reviewing Pull Request

**Pull Request**
- [x] Title follows correct format (Specifies Instant Answer + Purpose)
- [x] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Instant Answer Page** (for new Instant Answers)
- [x] Instant Answer page is correctly filled out and contains:
    - One topic for the Search Space Language (Java, Python, Scala, Ruby, etc.)
    - One topic from: Reference, Help, Libraries, Tools
        - Documentation Fatheads are considered "Reference"
    - Description, Data source, and 2+ example queries
    - Perl Module (e.g. "DDG::Fathead::PerlDoc" -- we only need a name, not an actual file)
    - Source Name (for "More at <source_name>" link)
    - Source Domain (must contain http:// or https:// -- can be the same as Data Source)
    - Source Info (used as Subtitle for each Article -- usually matches the IA Name)
    - 'Skip Abstract' is checked off
    - Source ID (ping @moollaza to assign one, once Fathead is ready for Beta deploy)

**Code**
- [x] Uniformly indented, well commented
- [ ] Fetch.sh and Parse.sh run without errors
- [ ] Output contains no blank lines, or multi-line entries
- [ ] Fathead Tests are passing (run `$ duckpan test <fathead_id>`)
    - Tester should report any failures

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/itext7

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: lib/fathead/itext7/data.url [hunk 1]
```diff
@@ -0,0 +1 @@
+http://itextsupport.com/apidocs/itext7/7.0.1/
```
Comment (pjhampton) [inline lib/fathead/itext7/data.url]: This isn't needed any more 👍  Please remove

File: lib/fathead/itext7/trigger_words.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+itext
+itext7
```
Comment (pjhampton) [inline lib/fathead/itext7/trigger_words.txt]: this entry on line 2 doesn't need to be there.

**Comment by BillyBarbaro** [comment]
Fixed the above comments.  I tried to take a look at it on Beta, but it seemed like it wasn't triggering.  Any advice on how to look into this?

**Comment by moollaza** [comment]
> Any advice on how to look into this?

@BillyBarbaro this is a great start! It looks like about 3000 of the articles contain `<div>` tags, and others contain `<a>` tags -- currently our backend ignore any articles with those in them.

Can you please remove any `<div>` tags in the abstracts, and replace any `<a>` tags with just their text?

**Comment by BillyBarbaro** [comment]
@moollaza Thanks for the feedback!  Sorry I missed those points.  The divs and anchor tags should be removed now, although I'm still not seeing my IA when search on beta.

**Comment by pjhampton** [comment]
Hey @BillyBarbaro

It seems that some entries aren't valid, like this one:

```
AbstractCMap    A                                       <section class="prog__container"></p></p></section> http://itextsupport.com/apidocs/itext7/7.0.1/com/itextpdf/io/font/cmap/AbstractCMap.html
```

You may need to fix this issue with the entries in output.txt before it will work on beta.

**Comment by BillyBarbaro** [comment]
@pjhampton Thanks for the help.  Looks like it's displaying now!

Those blank entries have no description in the documentation.  However, I thought it might still be useful to users to have the fathead show up so it could bring them to the class's documentation (in case they wanted to look up a constructor, method, etc).  Is there any way I can still do this or does this go against current convention?

**Comment by pjhampton** [comment]
Hey @BillyBarbaro, in the case of no description we'll just discard the entry! I agree the link would be helpful but no content is going to make a relatively poor IA.

**Comment by BillyBarbaro** [comment]
@moollaza I've made the requested changes and I can see my results on beta now.  This is ready for another round of review.

File: lib/fathead/itext7/trigger_words.txt [hunk 3]
```diff
@@ -0,0 +1,2 @@
+itext
+itext 7
```
Comment (moollaza) [inline lib/fathead/itext7/trigger_words.txt]: What about `itext7`?

**Comment by daxtheduck** [comment]
@BillyBarbaro this is now merged! It should be live on DuckDuckGo within one week (or during the next release) and we'll be sure to notify you when it happens.

After your Instant Answer goes live for the World to use, you will be able to see [traffic information](http://docs.duckduckhack.com/maintaining/tools.html#monitor-your-instant-answer-page) about your IA as well as receive feedback directly from users. 
                
Please refer to the [maintenance guide and docs](http://docs.duckduckhack.com/maintaining/guidelines.html) to keep up with potential changes to your Instant Answer.
                
Thank you again for your help!

File: lib/fathead/itext7/trigger_words.txt (hunk 3 above)
Comment (BillyBarbaro) [inline lib/fathead/itext7/trigger_words.txt]: I had included this originally, but above, @pjhampton said this wasn't needed.

Comment (pjhampton) [inline lib/fathead/itext7/trigger_words.txt]: Yep, sorry @BillyBarbaro. This was my mistake

**Pull request merged**